java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java

```diff
@@ -259,7 +259,8 @@ private static ArrowMessage frame(BufferAllocator allocator, final InputStream s
       ArrowBuf body = null;
       ArrowBuf appMetadata = null;
       while (stream.available() > 0) {
-        int tag = readRawVarint32(stream);
+        int tag = readRawVarint32WithEOFCheck(stream);
+
         switch (tag) {
```

Comment:
Can we explicitly handle -1 (EOF) here?

Reply:
That's not true, though, since readRawVarint32WithEOFCheck has a branch that returns -1.

Reply:
Ah, actually, now that I look at this, ArrowMessage#frame overall doesn't seem right...it shouldn't be checking stream.available() at all, instead it should be looping and reading the first byte, and handing it to readRawVarint32(int, InputStream) if it's >= 0.

I think that would explain the actual issues you are seeing - we should not rely on available() at all and instead use read(). So it should look like

```java
while (true) {
    int firstByte = stream.read();
    if (firstByte < 0) {
        // EOF
        break;    
    }
    int tag = readRawVarint32(firstByte, stream);
    // ...
}
```

Reply:
As-is, -1 will get silently ignored and that happens to be correct here, fortunately. But we shouldn't be relying on available() at all.

Reply:
Em,  it is similar to current code. Besides, current we can return unified exception information for abnormal EOF.

Reply:
The current code is wrong, though. (It would be entirely valid for available() == 0 even when the stream is not at EOF. The JavaDocs state that EOF implies available() == 0, but not the other way around.) And the PR raises different exceptions for EOF depending on the case.

Instead of adding flags to massage the current behavior in different cases, let's write code that adheres to the contract of InputStream.